### ISSUE TITLE
java.lang.IllegalStateException: Unexpected types in SetInRule.powSetIn

### ISSUE THREAD

**Issue by nano-o** [informalsystems/apalache#1824]
Hello, when running `apalache-mc check --inv=TypeOkay StarbridgeEthToStellar.tla` (spec is attached in zip file) I get
```
PASS #12: BoundedChecker                                          I@17:22:58.537                                                                                                                                                              
State 0: Checking 14 state invariants                             I@17:22:59.113                                                                                                                                                              
(Please report an issue at: [https://github.com/informalsystems/apalache/issues],java.lang.IllegalStateException: Unexpected types in SetInRule.powSetIn: expectedType = [amount: Int, from: ETH_ACCNT, hash: HASH, to: ETH_ACCNT], actualType
 = [amount: Int, from: ETH_ACCNT, h: HASH, to: ETH_ACCNT])
```
I can see my types are wrong but Apalache probably should report a type error instead of crashing.
[spec.zip](https://github.com/informalsystems/apalache/files/8773261/spec.zip)

**Comment by shonfeder** [comment]
Thanks for the report! 

In general, type checking of records will be improved by ongoing work tracked in https://github.com/informalsystems/apalache/issues/401, but this shouldn't be raising a state exception in any case!

**Comment by shonfeder** [comment]
Between the pending work on the typing for records and the work on the typed builder, I don't think it's worth trying to patch this problem at the moment. Seems to be a low impact temporary fix.

Will discuss tomorrow tho.

**Comment by konnov** [comment]
Hey @nano-o!

We have just finished working on precise type checking for records and variants, see the [HOWTO](https://apalache.informal.systems/docs/HOWTOs/howto-write-type-annotations.html#recipe-4-using-variants-in-heterogenous-sets). I wonder, whether the bug persists, if you transition to the new type system. I hope not :)

**Comment by Kukovec** [comment]
I've tested this today, adding in new-style record type annotations. I had to actually fix 1 type error:
The original spec defines
```
ETH_TX = [from : ETH_ACCNT, to : ETH_ACCNT, amount : Int, h : HASH] 
```
which I've changed to 
```
ethTx = {from : ETH_ACCNT, to : ETH_ACCNT, amount : Int, hash : HASH}
```
note that the original type alias defines a field `h`, _but_ in `Ethereum!ReceiveTx`, there is a line `mempool' = mempool \union {r}`, where `r` is a record defining a field `hash`, and not `h`, which originally caused the type checker to fail there. After fixing the alias, I got
```
The outcome is: NoError                                           
Checker reports no error up to computation length 5 
```
( I only ran 5 steps, for brevity).

I'm closing this issue, as using the new record types works as intended for this spec.
